poetry/installation/authenticator.py

```diff
@@ -0,0 +1,133 @@
+    def __init__(self, config, io):  # type: (Config, IO) -> None
+        self._config = config
+        self._io = io
+        self._session = None
+        self._credentials = {}
+        self._password_manager = PasswordManager(self._config)
+
+    @property
+    def session(self):  # type: () -> Session
+        from requests import Session  # noqa
+
+        if self._session is None:
+            self._session = Session()
+
+        return self._session
+
+    def request(self, method, url, **kwargs):  # type: (str, str, Any) -> Response
+        from requests import Request  # noqa
+        from requests.auth import HTTPBasicAuth
+
+        request = Request(method, url)
+
+        username, password = self._get_credentials_for_url(url)
+
+        if username is not None and password is not None:
+            request = HTTPBasicAuth(username, password)(request)
+
+        session = self.session
+        prepared_request = session.prepare_request(request)
+
+        proxies = kwargs.get("proxies", {})
+        stream = kwargs.get("stream")
+        verify = kwargs.get("verify")
+        cert = kwargs.get("cert")
+
+        settings = session.merge_environment_settings(
+            prepared_request.url, proxies, stream, verify, cert
+        )
+
+        # Send the request.
+        send_kwargs = {
+            "timeout": kwargs.get("timeout"),
+            "allow_redirects": kwargs.get("allow_redirects", True),
+        }
+        send_kwargs.update(settings)
+        resp = session.send(prepared_request, **send_kwargs)
+
+        resp.raise_for_status()
+
+        return resp
+
+    def _get_credentials_for_url(
+        self, url
+    ):  # type: (str) -> Tuple[Optional[str], Optional[str]]
+        parsed_url = urlparse.urlsplit(url)
+
+        netloc = parsed_url.netloc
+
+        credentials = self._credentials.get(netloc, (None, None))
+
+        if credentials == (None, None):
+            if "@" not in netloc:
+                credentials = self._get_credentials_for_netloc_from_config(netloc)
+            else:
+                # Split from the right because that's how urllib.parse.urlsplit()
+                # behaves if more than one @ is present (which can be checked using
+                # the password attribute of urlsplit()'s return value).
+                auth, netloc = netloc.rsplit("@", 1)
+                if ":" in auth:
+                    # Split from the left because that's how urllib.parse.urlsplit()
+                    # behaves if more than one : is present (which again can be checked
+                    # using the password attribute of the return value)
+                    credentials = auth.split(":", 1)
+                else:
+                    credentials = auth, None
+
+                credentials = tuple(
+                    None if x is None else urlparse.unquote(x) for x in credentials
+                )
+
+        if credentials[0] is not None or credentials[0] is not None:
```

Comment:
seems like this is just the same thing twice? I suspect the second one is meant to be `credentials[1]`?

Reply:
Nice catch!

I just fixed it and added tests accordingly.